File: csharp/DVSClient/DVSClient.Tests/phone/PhoneClientTests.cs
using DVSClient.Exceptions;
using DVSClient.Phone.Validate;
using DVSClientTests;
using NUnit.Framework;

namespace DVSClient.Phone.Tests
{
    [TestFixture]
    public class PhoneClientTests
    {
        [OneTimeSetUp]
        public void TestSetup()
        {
            Setup.LoadEnv();
        }
        
        [Test]
        public void Authentication_TokenNotSupplied_Throws()
        {
            var ex = Assert.Throws<InvalidConfigurationException>(() => PhoneConfiguration.NewBuilder("").Build());
            Assert.That(ex?.Message == "The supplied configuration must contain an authorisation token.");

            ex = Assert.Throws<InvalidConfigurationException>(() => PhoneConfiguration.NewBuilder(null).Build());
            Assert.That(ex?.Message == "The supplied configuration must contain an authorisation token.");
        }

        [Test]
        public void Authentication_InvalidTokenSupplied_Throws()
        {
            var token = "ThisIsNotAValidToken";
            var configuration = PhoneConfiguration.NewBuilder(token).Build();
            var client = ExperianDataValidation.GetPhoneClient(configuration);

            var ex = Assert.Throws<UnauthorizedException>(() => client.ValidateAsync("+44123456767", Setup.GetUniqueReferenceId()).GetAwaiter().GetResult());
            Assert.That(ex?.Message == "The authentication token you've provided is incorrect. Please check the Self Service Portal to find the right token.");
        }

        [Test]
        public void Validate_PhoneNumber_WithOutputFormat()
        {
            var configuration = PhoneConfiguration
                .NewBuilder(Setup.ValidTokenPhone)
                .UseOutputFormat("NATIONAL")
                .Build();
            var client = ExperianDataValidation.GetPhoneClient(configuration);
            var result = client.ValidateAsync("+442074987788", Setup.GetUniqueReferenceId()).GetAwaiter().GetResult();

            Assert.That(result.Confidence, Is.EqualTo(PhoneConfidence.NoCoverage));
            Assert.That(result.PhoneType, Is.EqualTo(PhoneType.Landline));
            Assert.That(result.FormattedPhoneNumber, Is.EqualTo("020 7498 7788"));
        }

        [Test]
        public void Validate_PhoneNumber_WithMetadata()
        {
            var configuration = PhoneConfiguration
                .NewBuilder(Setup.ValidTokenPhone)
                .IncludeMetadata()
                .Build();
            var client = ExperianDataValidation.GetPhoneClient(configuration);
            var result = client.ValidateAsync("+442074987788", Setup.GetUniqueReferenceId()).GetAwaiter().GetResult();

            Assert.That(result.Confidence, Is.EqualTo(PhoneConfidence.NoCoverage));
            Assert.That(result.PhoneType, Is.EqualTo(PhoneType.Landline));
            Assert.That(result.Metadata, Is.Not.Null);
            Assert.That(result.Metadata?.PhoneDetail, Is.Not.Null);
            Assert.That(result.Metadata?.PhoneDetail?.OriginalOperatorName, Is.EqualTo("BT"));
        }

        [Test]
        public void ReferenceId_OnBuilderStillWorks()
        {
            // Set reference ID on builder. Method is deprecated, but still provided for backwards compatibility
            var configuration = PhoneConfiguration.NewBuilder(Setup.ValidTokenPhone)
                .SetTransactionId(Setup.StaticReferenceId)
                .Build();
            var emailClient = ExperianDataValidation.GetPhoneClient(configuration);
            var result = emailClient.ValidateAsync("+442074987788").GetAwaiter().GetResult();

            Assert.That(result, Is.Not.Null);
            Assert.That(result.ReferenceId, Is.EqualTo(Setup.StaticReferenceId));
        }

        [Test]
        public void ReferenceId_OnMethodTakesPrecedence()
        {
            // Set reference ID on builder. Method is deprecated, but still provided for backwards compatibility
            // Setting the reference ID on the method should take precedence
            var configuration = PhoneConfiguration.NewBuilder(Setup.ValidTokenPhone)
                .SetTransactionId(Setup.GetUniqueReferenceId())
                .Build();
            var emailClient = ExperianDataValidation.GetPhoneClient(configuration);
            var result = emailClient.ValidateAsync("+442074987788", Setup.StaticReferenceId).GetAwaiter().GetResult();

            Assert.That(result, Is.Not.Null);
            Assert.That(result.ReferenceId, Is.EqualTo(Setup.StaticReferenceId));
        }

        [Test]
        public void ReferenceId_OnMethod()
        {
            var configuration = PhoneConfiguration.NewBuilder(Setup.ValidTokenPhone)
                .Build();
            var emailClient = ExperianDataValidation.GetPhoneClient(configuration);
            var result = emailClient.ValidateAsync("+442074987788", Setup.StaticReferenceId).GetAwaiter().GetResult();

            Assert.That(result, Is.Not.Null);
            Assert.That(result.ReferenceId, Is.EqualTo(Setup.StaticReferenceId));
        }

        [Test]
        public void ReferenceId_NotValueSpecified_UsesRandomValue()
        {
            var configuration = PhoneConfiguration.NewBuilder(Setup.ValidTokenPhone)
                 .Build();
            var emailClient = ExperianDataValidation.GetPhoneClient(configuration);
            var result = emailClient.ValidateAsync("+442074987788").GetAwaiter().GetResult();

            Assert.That(result, Is.Not.Null);
            Assert.That(result.ReferenceId, Is.Not.Empty);
            Assert.That(result.ReferenceId, Is.Not.EqualTo(Setup.StaticReferenceId));
        }
    }
}

File: java/src/test/java/com/experian/dvs/client/phone/PhoneClientTests.java
package com.experian.dvs.client.phone;

import com.experian.dvs.client.ExperianDataValidation;
import com.experian.dvs.client.Setup;
import com.experian.dvs.client.exceptions.InvalidConfigurationException;
import com.experian.dvs.client.exceptions.UnauthorizedException;
import com.experian.dvs.client.phone.validate.PhoneType;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class PhoneClientTests {
    @BeforeAll
    public static void setup() {
        Setup.loadEnv();
    }

    @Test
    public void testValidTokenPhone() {
        assertThat(Setup.VALID_TOKEN_PHONE).isNotEmpty();
    }

    @Test
    void authentication_TokenNotSupplied_Throws() {
        final Exception ex1 =  assertThrows(InvalidConfigurationException.class, () -> PhoneConfiguration.newBuilder("").build());
        assertThat(ex1.getMessage()).isEqualTo("The supplied configuration must contain an authorisation token");

        final Exception ex2 =  assertThrows(InvalidConfigurationException.class, () -> PhoneConfiguration.newBuilder(null).build());
        assertThat(ex2.getMessage()).isEqualTo("The supplied configuration must contain an authorisation token");
    }

    @Test
    void authentication_InvalidTokenSupplied_Throws() {
        final String token = "ThisIsNotAValidToken";
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(token)
                .build();

        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        final Exception ex = assertThrows(UnauthorizedException.class, () -> client.validate("+44123456767", Setup.getUniqueReferenceId()));
        assertThat(ex.getMessage()).isEqualTo("The authentication token you've provided is incorrect. Please check the Self Service Portal to find the right token.");
    }

    @Test
    void validate_PhoneNumber_WithOutputFormat() {
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(Setup.VALID_TOKEN_PHONE)
                .useOutputFormat("NATIONAL")
                .build();
        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        final var result = client.validate("+442074987788", Setup.getUniqueReferenceId());

        assertThat(result.getConfidence()).isEqualTo(PhoneConfidence.NO_COVERAGE);
        assertThat(result.getPhoneType()).isEqualTo(PhoneType.LANDLINE);
        assertThat(result.getFormattedPhoneNumber()).isEqualTo("020 7498 7788");
    }

    @Test
    void Validate_PhoneNumber_WithMetadata() {
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(Setup.VALID_TOKEN_PHONE)
                .includeMetadata()
                .build();
        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        final var result = client.validate("+442074987788", Setup.getUniqueReferenceId());

        assertThat(result.getConfidence()).isEqualTo(PhoneConfidence.NO_COVERAGE);
        assertThat(result.getPhoneType()).isEqualTo(PhoneType.LANDLINE);
        assertThat(result.getMetadata()).isNotNull();
        assertThat(result.getMetadata().getPhoneDetail()).isNotNull();
        assertThat(result.getMetadata().getPhoneDetail().getOriginalOperatorName()).isEqualTo("BT");
    }

    @Test
    void referenceId_OnBuilderStillWorks() {
        // Set reference ID on builder. Method is deprecated, but still provided for backwards compatibility
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(Setup.VALID_TOKEN_PHONE)
                .setTransactionId("specialRefId")
                .build();
        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        var result = client.validate("+442074987788");

        assertThat(result).isNotNull();
        assertThat(result.getReferenceId()).isEqualTo("specialRefId");
    }

    @Test
    void referenceId_OnMethodTakesPrecedence() {
        // Set reference ID on builder. Method is deprecated, but still provided for backwards compatibility
        // Setting the reference ID on the method should take precedence
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(Setup.VALID_TOKEN_PHONE)
                .setTransactionId(Setup.getUniqueReferenceId())
                .build();
        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        var result = client.validate("+442074987788", "specialRefId");

        assertThat(result).isNotNull();
        assertThat(result.getReferenceId()).isEqualTo("specialRefId");
    }

    @Test
    void referenceId_OnMethod() {
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(Setup.VALID_TOKEN_PHONE)
                .build();
        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        var result = client.validate("+442074987788", "specialRefId");

        assertThat(result).isNotNull();
        assertThat(result.getReferenceId()).isEqualTo("specialRefId");
    }

    @Test
    void referenceId_NotValueSpecified_UsesRandomValue() {
        final PhoneConfiguration configuration = PhoneConfiguration
                .newBuilder(Setup.VALID_TOKEN_PHONE)
                .build();
        final PhoneClient client = ExperianDataValidation.getPhoneClient(configuration);
        var result = client.validate("+442074987788");

        assertThat(result).isNotNull();
        assertThat(result.getReferenceId()).isNotEmpty();
    }
}
